pvlib/ivtools.py

```diff
@@ -0,0 +1,334 @@
+
+    .. math::
+
+        I ~ IL/(1+Gp*Rs) - (Gp*V)/(1+Gp*Rs) = beta0 + beta1*V
+
+        3) fit the exponential portion of the IV curve defined by
+            beta0 + beta1*V - I > ilim * i_sc, where
+            exp((V+I*Rs)/(nNsVth)) >> 1 so that
+            exp((V+I*Rs)/(nNsVth)) - 1 ≈ exp((V+I*Rs)/(nNsVth))
+
+    .. math::
+
+        log(beta0 - beta1*V - I) ~ log((I0)/(1+Gp*Rs)) + (V)/(nNsVth) +
+        (Rs*I)/(nNsVth) = beta2 + beta3*V + beta4*I
+
+        4) calculate values for ``IL, I0, Rs, Rsh,`` and ``nNsVth`` from the
+            regression coefficents beta0, beta1, beta3 and beta4.
+
+
+    References
+    ----------
+    [1] S.R. Wenham, M.A. Green, M.E. Watt, "Applied Photovoltaics" ISBN
+    0 86758 909 4
+    [2] C. B. Jones, C. W. Hansen, Single Diode Parameter Extraction from
+    In-Field Photovoltaic I-V Curves on a Single Board Computer, 46th IEEE
+    Photovoltaic Specialist Conference, Chicago, IL, 2019
+    """
+
+    # If not provided, extract v_oc, i_sc, v_mp and i_mp from the IV curve data
+    if v_oc is None:
+        v_oc = v[-1]
+    if i_sc is None:
+        i_sc = i[0]
+    if mp is not None:
+        v_mp, i_mp = mp
+    else:
+        v_mp, i_mp = _find_mp(v, i)
+
+    # Find beta0 and beta1 from linear portion of the IV curve
+    beta0, beta1 = _find_beta0_beta1(v, i, vlim, v_oc)
+
+    if not np.isnan(beta0):
+        beta3, beta4 = _find_beta3_beta4(v, i, beta0, beta1, ilim, i_sc)
+
+    # calculate single diode parameters from regression coefficients
+    result = _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp,
+                                       v_oc)
+
+    if result is None:
+        raise RuntimeError("Parameter extraction failed. Try increasing the"
+                           "number of data points in the IV curve")
+    else:
+        IL, I0, Rsh, Rs, nNsVth = result
+        return IL, I0, Rsh, Rs, nNsVth
+
+
+def _find_mp(v, i):
+    """
+    Finds voltage and current at maximum power point.
+
+    Parameters
+    ----------
+    v : ndarray
+        1D array containing voltage at each point on the IV curve, increasing
+        from 0 to v_oc inclusive, of `float` type [V]
+
+    i : ndarray
+        1D array containing current at each point on the IV curve, decreasing
+        from i_sc to 0 inclusive, of `float` type [A]
+
+    Returns
+    -------
+    v, i : tuple
+        voltage ``v`` and current ``i`` at the maximum power point [V], [A]
+    """
+    p = v * i
+    idx = np.argmax(p)
+    return v[idx], i[idx]
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
+
+
+def _find_beta0_beta1(v, i, vlim, v_oc):
+    # Get intercept and slope of linear portion of IV curve.
+    # Start with V =< vlim * v_oc, extend by adding points until slope is
+    # negative (downward).
+    beta0 = np.nan
+    beta1 = np.nan
+    idx = np.searchsorted(v, vlim * v_oc)
+    while idx <= len(v):
```

Comment:
- Redundant `len(v)` calculated each loop
- suggest either instead define `numpoints = len(v)` on line before loop
- or even better just use `try: except IndexError` might perform best?

```python
    try:
        coef = np.polyfit(v[:idx], i[:idx], deg=1)
        if coef[0] < 0:
            # intercept term
            beta0 = coef[1].item()
            # sign change of slope to get positive parameter value
            beta1 = -coef[0].item()
            break
        else:
            idx += 1
    except IndexError as e:
        raise RuntimeError("no fit")

Reply:
Also, since you use `break`, a good pattern might be a for loop

    for idx in len(v):
        coef = np.polyfit(v[:idx], i[:idx], deg=1)
        if coef[0] < 0:
            # intercept term
            beta0 = coef[1].item()
            # sign change of slope to get positive parameter value
            beta1 = -coef[0].item()
            break